File: go/server/executor_selection_test.go
package server

import (
	"context"
	"testing"

	core "github.com/SmooAI/smooth-operator-core/go/core"
)

// fakeExecutor is a stand-in durable AgentExecutor: it records whether it was
// driven, so a test can prove the selection actually routed the turn to it (not
// just returned it). It never runs a real turn.
type fakeExecutor struct {
	used bool
}

func (f *fakeExecutor) Execute(_ context.Context, _ *core.SmoothAgent, _ string, _ []core.ChatMessage) (core.AgentRunResponse, error) {
	f.used = true
	return core.AgentRunResponse{}, nil
}

func (f *fakeExecutor) ExecuteStreaming(_ context.Context, _ *core.SmoothAgent, _ string, _ *core.SmoothAgentThread) (*core.Stream, error) {
	f.used = true
	return nil, nil
}

// With durable mode opted in AND an executor injected, the injected durable
// executor is handed back verbatim — the SAME instance the turn will drive. This
// is the whole point of the injected slot: a durable backend built outside this
// binary has to survive the trip through turnExecutor.
func TestTurnExecutor_DurableOnUsesInjected(t *testing.T) {
	fake := &fakeExecutor{}
	for _, on := range []string{"1", "true", "TRUE", " on ", "yes"} {
		selected := turnExecutor(fake, on)
		if selected != core.AgentExecutor(fake) {
			t.Fatalf("env=%q: expected injected durable executor to be used, got %T", on, selected)
		}
	}
}

// Durable mode off (unset/empty/unrecognized) ignores the injected slot and runs
// in-process, even when a durable executor is present — getting this backwards
// would silently make every deployed turn durable.
func TestTurnExecutor_DurableOffFallsBackToInProcess(t *testing.T) {
	fake := &fakeExecutor{}
	for _, off := range []string{"", " ", "0", "false", "off", "no", "maybe"} {
		selected := turnExecutor(fake, off)
		if selected == core.AgentExecutor(fake) {
			t.Fatalf("env=%q: durable is opt-in only; injected executor must NOT be used", off)
		}
		if _, ok := selected.(*core.InProcessExecutor); !ok {
			t.Fatalf("env=%q: expected *core.InProcessExecutor, got %T", off, selected)
		}
	}
}

// Durable requested but nothing injected ⇒ a fresh in-process executor (warn +
// fall back), never a nil that would panic the turn.
func TestTurnExecutor_DurableRequestedButNothingInjected(t *testing.T) {
	selected := turnExecutor(nil, "1")
	if _, ok := selected.(*core.InProcessExecutor); !ok {
		t.Fatalf("expected fallback *core.InProcessExecutor, got %T", selected)
	}
}

// Nothing injected, durable off ⇒ a non-nil in-process executor. (Unlike Rust's
// Arc, Go pointers to the zero-size InProcessExecutor may compare equal, so the
// invariant here is the concrete type, not distinct identity.)
func TestTurnExecutor_NoInjectionBuildsInProcess(t *testing.T) {
	selected := turnExecutor(nil, "")
	if selected == nil {
		t.Fatal("expected a non-nil in-process executor")
	}
	if _, ok := selected.(*core.InProcessExecutor); !ok {
		t.Fatalf("expected *core.InProcessExecutor, got %T", selected)
	}
}

// The opt-in parse matches the Rust reference exactly.
func TestDurableRequested(t *testing.T) {
	for _, on := range []string{"1", "true", "TRUE", " on ", "yes", "Yes", "  1  "} {
		if !durableRequested(on) {
			t.Errorf("%q should opt in", on)
		}
	}
	for _, off := range []string{"", " ", "0", "false", "off", "no", "maybe", "enabled"} {
		if durableRequested(off) {
			t.Errorf("%q should stay off", off)
		}
	}
}

File: typescript/server/test/executor-selection.test.ts
/**
 * Tests for the durable-executor selection seam — the TS sibling of the Rust
 * server's `runner.rs` `durable_requested` / `turn_executor` tests.
 *
 * Getting this backwards would silently change how every deployed turn runs, so
 * the opt-in parse and the injection precedence are pinned here. Uses a FAKE
 * injected executor (no Temporal dependency) to prove: env on + injected ⇒ the
 * durable executor is used; nothing injected ⇒ the in-process default.
 */

import { InProcessExecutor } from '@smooai/smooth-operator-core';
import type { AgentExecutor, AgentRunResponse, SmoothAgent, StreamEvent } from '@smooai/smooth-operator-core';
import { afterEach, describe, expect, it, vi } from 'vitest';

import { DURABLE_EXECUTOR_ENV, durableRequested, turnExecutor } from '../src/executorSelection.js';

/** A stand-in durable backend — proves an injected executor survives the trip. */
class FakeExecutor implements AgentExecutor {
    async execute(): Promise<AgentRunResponse> {
        return { text: '', iterations: 0, toolCalls: 0, usage: { promptTokens: 0, completionTokens: 0 }, costUsd: 0, budgetExceeded: false };
    }

    async *executeStreaming(): AsyncGenerator<StreamEvent> {
        yield { type: 'done', response: await this.execute() };
    }
}

// The runner passes the agent + message; the seam never touches them, so an empty
// object is enough for these selection tests.
const noopAgent = {} as unknown as SmoothAgent;

afterEach(() => {
    vi.restoreAllMocks();
});

describe('durableRequested', () => {
    it('is opt-in only — on for 1/true/on/yes (any case/space), off for everything else', () => {
        for (const on of ['1', 'true', 'TRUE', ' on ', 'yes']) {
            expect(durableRequested(on)).toBe(true);
        }
        for (const off of ['', ' ', '0', 'false', 'off', 'no', 'maybe']) {
            expect(durableRequested(off)).toBe(false);
        }
        expect(durableRequested(undefined)).toBe(false);
    });
});

describe('turnExecutor', () => {
    it('uses an injected executor verbatim — the same handle, env not consulted', () => {
        const injected = new FakeExecutor();
        expect(turnExecutor(injected, { [DURABLE_EXECUTOR_ENV]: '1' })).toBe(injected);
        expect(turnExecutor(injected, {})).toBe(injected);
    });

    it('env on + injected ⇒ the durable executor is used', async () => {
        const injected = new FakeExecutor();
        const spy = vi.spyOn(injected, 'executeStreaming');
        const selected = turnExecutor(injected, { [DURABLE_EXECUTOR_ENV]: 'true' });
        // Drain the seam to prove the injected backend is what actually runs.
        for await (const _ of selected.executeStreaming(noopAgent, 'hi', [])) {
            // consume
        }
        expect(selected).toBe(injected);
        expect(spy).toHaveBeenCalledOnce();
    });

    it('env off + nothing injected ⇒ a fresh in-process executor', () => {
        const a = turnExecutor(undefined, {});
        const b = turnExecutor(undefined, {});
        expect(a).toBeInstanceOf(InProcessExecutor);
        expect(b).toBeInstanceOf(InProcessExecutor);
        expect(a).not.toBe(b); // each fallback builds its own
    });

    it('env on + nothing injected ⇒ warns and falls back to in-process (never a fake-durable turn)', () => {
        const warn = vi.spyOn(console, 'warn').mockImplementation(() => {});
        const selected = turnExecutor(undefined, { [DURABLE_EXECUTOR_ENV]: 'yes' });
        expect(selected).toBeInstanceOf(InProcessExecutor);
        expect(warn).toHaveBeenCalledOnce();
        expect(warn.mock.calls[0][0]).toContain(DURABLE_EXECUTOR_ENV);
    });
});
